Proposal: Remote API authorization

calavera opened the pull request:
This proposal describes how to use Google's [Macaroons](http://research.google.com/pubs/pub41892.html) as authorization layer for the remote api.

I haven't written any code yet, I'd like to get feedback to know if people would be interested in adding something like this to the core.

As a naive explanation, Macaroons are secure layered cookies where you can add extra context. For instance, you can generate a macaroon to give only read access to the api. On top of that layer, you could add another caveat to give read access only to a specific container. I hope you get the idea, but I'll be happy to answer any further question.

Signed-off-by: David Calavera david.calavera@gmail.com

calavera committed:
Add first draft of the remote api authorization.

Signed-off-by: David Calavera <david.calavera@gmail.com>

jfrazelle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

crosbymichael commented:
/me adds the paper to pocket....

SvenDowideit commented on docs/sources/articles/remote_api_authorization.md:
```diff
@@ -0,0 +1,63 @@
+
+## Sharing credentials
+
+You can always see a serialized version of your macaroon by running `id --show`.
```
`docker id --show`?

calavera replied:
yes!

SvenDowideit commented on docs/sources/articles/remote_api_authorization.md:
```diff
@@ -0,0 +1,63 @@
+
+## Initializing the credentials in the server
+
+The first command to run is `docker id --init`. This will create the root macaroon in the server and it will be shared with the client.
```
how and when will it be shared with the client?

calavera replied:
Definitely, I think this proposal still needs a more detailed part explaining how all this works from the client perspective.

SvenDowideit commented on docs/sources/articles/remote_api_authorization.md:
````diff
@@ -0,0 +1,63 @@
+This is the command to allow only read access to only one container:
+
+```
+$ docker id --create --secret "Super secret token" --op read --container-id "container id"
````
can you add an example of the client side request using this new secret?

calavera committed:
Add outputs to some examples.

- Move to env variables for a one time usage of a macaroon, that way we don't need to add an option to every command.

calavera commented:
@SvenDowideit I've added outputs to some of the commands and tried to explain better how to share macaroon with other clients.

I'm trying to avoid explaining internal functionality, but it might be good to talk about it.

Macaroons are shared between client and server via cookies. It's not much complicated than that, the client sends the macaroon in a cookie for authorization and the server validates it. If the client doesn't have access to operation returns a `403`.

Please, let me know if this is more clear, I really appreciate any feedback I can get on this.

icecrime commented:
Ping @diogomonica @NathanMcCauley!

NathanMcCauley commented on docs/sources/articles/remote_api_authorization.md:
````diff
@@ -0,0 +1,103 @@
+eyJjYXZlYXRzIjpbXSwibG9jYXRpb24iOiJ0aGUgY2xvdWQiLCJpZGVudGlmaWVyIjoiZG9ja2VyIiwic2lnbmF0dXJlIjoiNzE4ZGJkZTQwZDliZGFhNjM4MTUzOTQxZTA3MjY4YjlhYTkzYzk1ODIyMzVlMGI4ZWM0NDgzZGQ4ZmVkNTMyZiJ9
+```
+
+You can share your serialized macaroon with any other client that needs access to the docker api. There are two ways to tell the docker client to use a specific macaroon.
````
Could you provide an example of how sharing this macaroon with another client would work? Is this expected to be an out-of-band operation? Or integrated sharing through another mechanism?

NathanMcCauley replied:
The utility of this proposal entirely hinges on whether we have a strong use case for decentralized attenuation of capabilities.

NathanMcCauley commented on docs/sources/articles/remote_api_authorization.md:
```diff
@@ -0,0 +1,103 @@
+- op: Restrict access only to certain request operations. The only allowed values are `read` and `write`.
+- image_id: Restrict access only to one image. The value must be an image id present in the server.
+- container_id: Restrict access only to one container. The value must be a container id present in the server.
+- ip: Restrict access from only one IP. The value must be a value IPv4 or IPv6.
```
Would the server be expected to check the IP matches in the case of this caveat? Generally its preferable to build authn/authz at layer 7 rather than level 4.

NathanMcCauley commented on docs/sources/articles/remote_api_authorization.md:
```diff
@@ -0,0 +1,103 @@
+- image_id: Restrict access only to one image. The value must be an image id present in the server.
+- container_id: Restrict access only to one container. The value must be a container id present in the server.
+- ip: Restrict access from only one IP. The value must be a value IPv4 or IPv6.
+- expires: Restrict access for a period of time. The value must be the allowed interval in seconds. For instance, `60` to allow access for only 60 seconds.
```
Are caveats timestamped? How does the server check validity period?

NathanMcCauley commented on docs/sources/articles/remote_api_authorization.md:
```diff
@@ -0,0 +1,103 @@
+
+Using the root macaroon, you can create new macaroons with caveats to share with other people. Caveats are used to restrict access to certain parts of the server.
+
+These are the current allowed caveats:
```
It might make sense to generalize the caveat system by specifying allowed/disallowed resources (by regex or wildcarding) from the API.

Some examples might be

```
ALLOW GET  /containers/*/
DISALLOW POST /containers/*/stop
ALLOW POST /build/
```

This might be too much flexibility -- it might be better to just specify a number of high level role definitions and assign particular abilities to those named roles. Roles could be `admin`, `logs`, `lifecycle`, etc. @calavera WDYT?

calavera commented:
I'm going to close this PR for now. I'll address those comments when I have a more clear path to a good implementation.

Thank you all for your feedback!
